src/python/WMCore/Services/Rucio/Rucio.py

```diff
@@ -478,10 +480,79 @@ def createReplicationRule(self, names, rseExpression, scope='cms', copies=1, **k
         except AccessDenied as ex:
             msg = "AccessDenied creating DID replication rule. Error: %s" % str(ex)
             raise WMRucioException(msg)
+        except DuplicateRule as ex:
+            # NOTE:
+            #    The unique constraint is per did and it is checked against the tuple:
+            #    rucioAccount + did = (name, scope) + rseExpression
+            # NOTE:
+            #    This exception will be thrown by Rucio even if a single Did has
+            #    a duplicate rule. In this case all the rest of the Dids will be
+            #    ignored, which in general should be addressed by Rucio. But since
+            #    it is not, we should break the list of Dids and proceed one by one
+            # NOTE:
+            #    This thing here may be slow, because it will wait for Rucio to
+            #    return the history of rules per every Did, but no shorter path exists
+            msg = "A duplicate rule for: \naccount: %s \ndids: %s \nrseExpression: %s\ncopies: %s\n"
+            for did in dids:
+                try:
+                    response = self.cli.add_replication_rule([did], copies, rseExpression, **kwargs)
```

Comment:
Don't we need to save all of the response rules id? Then we only list the rule history for those that really have a DuplicateRule.

Reply:
I do not think so. We will not benefit from this, only a bunch of `if` conditions would come into the code which would make it harder to read. If it turns out that we suffer too much performance we can think of doing so, but even then I foresee not that much of optimization at the end.

Reply:
Ok, again :) I made few gentle touches to the code and we have this optimization now. Coming with my next commit.